src/Altinn.Platform/Altinn.Platform.Authentication/Authentication/Repositories/OrganisationRepository.cs

```diff
@@ -17,101 +17,113 @@ namespace Altinn.Platform.Authentication.Repositories
         /// Instantiates the class.
         /// </summary>
+        /// <param name="httpClient">the http client</param>
+        /// <param name="memoryCache">the memory cache</param>
         /// <param name="logger">the logger</param>
         /// <param name="generalSettings">the settings which contains the url to the json organisation file</param>
-        public OrganisationRepository(ILogger<OrganisationRepository> logger, IOptions<GeneralSettings> generalSettings)
+        public OrganisationRepository(HttpClient httpClient, IMemoryCache memoryCache, ILogger<OrganisationRepository> logger, IOptions<GeneralSettings> generalSettings)
         {
-            this.logger = logger;
-            organisationListLocation = new Uri(generalSettings.Value.GetOrganisationRepositoryLocation);
+            _httpClient = httpClient;
+            _logger = logger;
+            _organisationListLocation = new Uri(generalSettings.Value.GetOrganisationRepositoryLocation);
+
+            _memoryCache = memoryCache;
+            _cacheEntryOptions = new MemoryCacheEntryOptions()
+                .SetPriority(CacheItemPriority.Normal)
+                .SetAbsoluteExpiration(new TimeSpan(1, 0, 0));
         }
 
         /// <inheritdoc/>
         public async Task<Organisation> GetOrganisationByOrgNumber(string orgNumber)
         {
-            await HarvestOrgsIfCacheIsMoreThanOneHourOld();
+            string cacheKey = $"org-{orgNumber}";
+            Organisation organisation;
 
-            return orgNumberToOrganisation.GetValueOrDefault(orgNumber, null);
+            if (!_memoryCache.TryGetValue(cacheKey, out organisation))
+            {
+                await HarvestOrgs();
+                _memoryCache.TryGetValue(cacheKey, out organisation);
+            }
+
+            return organisation;
         }
 
         /// <inheritdoc/>
         public async Task<Organisation> GetOrganisationByOrg(string org)
         {
-            await HarvestOrgsIfCacheIsMoreThanOneHourOld();
+            string cacheKey = $"org-{org}";
+            Organisation organisation;
+
+            if (!_memoryCache.TryGetValue(cacheKey, out organisation))
+            {
+                await HarvestOrgs();
+                _memoryCache.TryGetValue(cacheKey, out organisation);
+            }
 
-            return orgToOrganisation.GetValueOrDefault(org, null);
+            return organisation;
         }
 
         /// <inheritdoc/>
         public async Task<string> LookupOrg(string orgNumber)
         {
-            await HarvestOrgsIfCacheIsMoreThanOneHourOld();
+            string cacheKey = $"org-{orgNumber}";
+            Organisation organisation;
 
-            Organisation organisation = orgNumberToOrganisation.GetValueOrDefault(orgNumber, null);
+            if (!_memoryCache.TryGetValue(cacheKey, out organisation))
+            {
+                await HarvestOrgs();
+                _memoryCache.TryGetValue(cacheKey, out organisation);
+            }
```

Comment:
This implementation will cause us to download the list everytime somone ask for an org not in the list. Maybe we instead should cache the original dictionary? That way we reuse the cache for at least 1 hour before updating.

Reply:
Caching the dictionary would mean having to create a second dictionary so we can easily lookup based on orgNo as well. Tried to address the problem by adding a dateTime parameter to keep track of last harvest, without modifying what is cached. 🍂